### DISCUSSION TITLE
Data List: Custom Value Converters

### DISCUSSION THREAD

**Discussion by leekelleher** [leekelleher/umbraco-contentment#89] [Idea]
> _Opening an issue to follow on from a discussion with @abjerner on commit 6dd0b4d4b31aa402d5d610c42e4b244114d328df._

### Is your feature request related to a problem? Please describe.

@greystate asked on the forum if it was possible for the built-in XML data source could return additional meta-data from the selected value? _e.g. the `Name`, `Icon`, etc._

<https://our.umbraco.com/packages/backoffice-extensions/contentment/contentment-feedback/102742-how-can-i-output-the-label-of-a-datalist-item>

Currently only the `Value` is stored against the property, and that's the value returned on the front-end. However we do have an interface `IDataListSourceValueConverter` for enabling custom object-types and values, e.g. for Enums _(and in upcoming v1.1 an `IPublishedContent` for an Umbraco content data source)._

I wonder if there is a better way to configure the type of value you would to be returned from a Data List?

### Describe the solution you'd like

Currently the `IDataListSourceValueConverter` interface can be used on a data source to convert the value to a specific type. However this ties the data source and output object-type close together, which irks me a little.  I'm wondering about the scenario where users could specify a different value type, e.g. for the Umbraco content data source, whether they would prefer to have the `IPublishedContent`, or the `Udi`, or the raw string of the UDI, or even the `int`?

On the data type configuration, we could have a 3rd field for "Value converter", with options on how to configure the desired output.  The default option would still be a `string` (and whether that'd be single or enumerable, depending on the list editor used).

I haven't considered much on the technical implementation of this yet, I envisage an interface (much like the data source and list editor interfaces), e.g. `IDataListValueConverter`, which would open it up to user customisation too.

**Comment by abjerner** [comment]
This is a step up from what I suggested, but I like it. This should developers the most amount of flexibility 😎 

I'd imagine it could work much like the `IDataListSourceValueConverter`, but with a few differences. Thinking out loud, the new interface could still describe a `GetValueType` method, but instead of a `Dictionary<string, object> config` parameter, it could have parameter with the data source like `IDataListSource source` (assuming you could get the data source configuration from that as well.

And the `ConvertValue` method could have an additional `IDataListSource source` parameter as well. Then depending on the requirements and implementation of the converter, it could call `source.GetItems(...)` to get the items if it needs to, or look up the IDs/values it self - eg. for `IPublishedContent`.

It could possibly also make sense to pass the `DataListDataEditor` instance along as well, as it then would be possible to access the overall configuration.

**Comment by leekelleher** [comment]
> _Work in progress update._

For those following along at home, I've done a proof of concept. See commit 6eb9653416d906b1ee26c0191a005678a5e9be58 for initial `IDataListValueConverter` code, etc.

So far I've prototyped 3 value converters, one for the Enum, another for `IPublishedContent` and a generic one for converting to standard .NET types (`string`, `int`, `Guid`, etc).

![Annotation 2020-07-02 164528](https://user-images.githubusercontent.com/209066/86380156-aef12b80-bc83-11ea-990e-4f980a23877e.png)

Ambitiously called "Universal Type Converter"... 😆 

![Annotation 2020-07-02 164613](https://user-images.githubusercontent.com/209066/86380159-af89c200-bc83-11ea-9cc8-2af0a7ee8ed6.png)

The idea is that the output value type can be configured in the backoffice without getting into C# code.

Notes _(for self),_ this would be a breaking change, as it removes the `IDataListSourceValueConverter` and introduces `IDataListValueConverter`, so I've put this in a **v2.0 branch**.  I'm also not convinced about having to manually configure value converters for data sources such as Enum. It was quite nice that it was automatically done for you (in v1.0). Would be nice to have this auto-wired up. 🤔

**Comment by greystate** [comment]
@leekelleher Just want to say that this looks utterly awesome Lee!

I'd have thunk this was a giant can of worms, to be honest :D  - I was really just thinking of how the Color Picker can be configured to either just return the hexcolor, or to return an object (tuple?) with `.Color` and `.Label` properties.

This would be all sorts of cool! 

/Chriztian

**Comment by bjarnef** [comment]
@leekelleher I haven't tried this package yet, but I wonder if it only makes sense to convert to `IPublishedContent` and if it should have an option for `IPublishedElement` as well?

**Comment by leekelleher** [comment]
Thanks @bjarnef. This is purely a prototype at the moment - nothing set in stone. 😃
Although in the case of converting values from the Umbraco Content data source, it'd only be `IPublishedContent`, as `IPublishedElement` isn't something that is _query-able_ with Umbraco's APIs, _(yet)._ But if that day may come, it'd be totally possible to do `IPublishedElement` too with this approach. 👍

**Comment by skttl** [comment]
Wouldn't it be possible to make a data list of elements from a block list / nested content, and then need a conversion to IPublishedElements?

**Comment by leekelleher** [comment]
@skttl Yeah, that'd be totally possible. I was only meaning that the "Umbraco Content" data source (which I'm developing for v1.1) would only return `IPublishedContent` objects.

It's all still in the prototyping phase, so anything is possible! 😃

**Comment by leekelleher** [comment]
Moved this topic to a discussion, as I'm not convinced if it's a good feature for Data List or not. I'm still having niggles about users needing to manually configure value converters for certain data sources. Totally open for discussion though. 👍

**Comment by leekelleher** [comment]
Closing this off, as it's not something I've needed to re-explore since the original brainstorm.
I did try rolling it out as a standalone package: [**Custom Value Converter**](https://github.com/leekelleher/umbraco-custom-valueconverter), but that hardly got any interest.
